README.md
# angular-stdlib
Angular standard library open source part of the *PACK Solutions* `@ps-angular-common` library

| Package | npm version |
| ------- | -------------------- |
| `@angular-stdlib/date` | [![npm version](https://badge.fury.io/js/%40angular-stdlib%2Fdate.svg)](https://badge.fury.io/js/%40angular-stdlib%2Fdate) |


CHANGELOG.md
# 0.3.2 (2021-05-17)

### Bug Fixes

* **deps-dev:** bump ng-packagr from 4.2.0 to 5.7.1 dependencies, closes ([#31](https://github.com/PACK-Solutions/angular-stdlib/issues/31))
* **deps:** bump yargs-parser from 5.0.0 to 5.0.1 dependencies, closes ([#35](https://github.com/PACK-Solutions/angular-stdlib/issues/35))
* **deps:** bump hosted-git-info from 2.8.5 to 2.8.9 dependencies, closes ([#34](https://github.com/PACK-Solutions/angular-stdlib/issues/34))
* **deps:** bump lodash from 4.17.15 to 4.17.21 dependencies, closes ([#33](https://github.com/PACK-Solutions/angular-stdlib/issues/33))
* **deps:** bump url-parse from 1.4.7 to 1.5.1 dependencies, closes ([#32](https://github.com/PACK-Solutions/angular-stdlib/issues/32))
* **deps:** bump handlebars from 4.5.3 to 4.7.7 dependencies, closes ([#30](https://github.com/PACK-Solutions/angular-stdlib/issues/30))
* **deps:** bump y18n from 3.2.1 to 3.2.2 dependencies, closes ([#29](https://github.com/PACK-Solutions/angular-stdlib/issues/29))
* **deps:** bump elliptic from 6.5.2 to 6.5.4 dependencies, closes ([#28](https://github.com/PACK-Solutions/angular-stdlib/issues/28))
* **deps:** bump http-proxy from 1.18.0 to 1.18.1 dependencies, closes ([#27](https://github.com/PACK-Solutions/angular-stdlib/issues/27))
* **deps:** bump websocket-extensions from 0.1.3 to 0.1.4 dependencies, closes ([#24](https://github.com/PACK-Solutions/angular-stdlib/issues/24))


# 0.3.1 (2020-01-22)

### GitHub

* add a contributing file, closes ([#9](https://github.com/PACK-Solutions/angular-stdlib/issues/9))
* **date:** add relevant keywords to `package.json`, closes ([#11](https://github.com/PACK-Solutions/angular-stdlib/issues/11))
* replace contact email address with oss@pack-solutions.com, closes ([#8](https://github.com/PACK-Solutions/angular-stdlib/issues/8))


# 0.3.0 (2020-01-04)

### BREAKING CHANGES

* **date:** `AngularDateHttpInterceptor` has been renamed `DateHttpInterceptor`, closes ([#4](https://github.com/PACK-Solutions/angular-stdlib/issues/4))

### Features

* **date:** add `DateValidators` built-in validators for Dates, closes ([#3](https://github.com/PACK-Solutions/angular-stdlib/issues/3))

### GitHub

* add a code of conduct (based on [Contributor Covenant Code of Conduct v1.4](https://www.contributor-covenant.org/version/1/4/code-of-conduct.html)), closes ([#6](https://github.com/PACK-Solutions/angular-stdlib/issues/6))


# 0.2.1 (2019-12-22)

### Documentation

* **date:** fix `AngularDateHttpInterceptor` example on README, closes ([#1](https://github.com/PACK-Solutions/angular-stdlib/issues/1))


# 0.2.0 (2019-11-18)

### Features

* **date:** add `DateFormatter.toFakeSerializedDate()` method that transforms a Date to have zero UTC offset timezone at serialization.


# 0.1.0 (2019-11-07)

### Features

* **date:** `AngularDateHttpInterceptor` provides an interceptor that intercepts HTTP calls and converts from JSON response body : **ISO 8601** date and datetime strings to native Date object.
  > NOTE - Only necessary for projects using dependencies like `ngx-bootstrap/datepicker`, this one does not handle *ISO string dates* unlike **Angular**.<br/>
  > Linked to this open issue : https://github.com/valor-software/ngx-bootstrap/issues/4487
* **date:** add `DateFormatter.ISO_DATE` formatter.


projects/angular-stdlib/date/src/lib/date-formatter.spec.ts
import { DateFormatter } from './date-formatter';

describe('DateFormatter', () => {

  it('should create an instance', () => {
    expect(new DateFormatter()).toBeTruthy();
  });

  it('should has ISO_DATE formatter', () => {
    expect(DateFormatter.ISO_DATE).toBe('yyyy-MM-dd');
  });

  it('should transform null date to fake serialized one', () => {
    expect(DateFormatter.toFakeSerializedDate(null)).toBeNull();
  });

  it('should transform timeless date to fake serialized one', () => {
    const fakeSerializedDate = DateFormatter.toFakeSerializedDate(new Date('2019-03-24T00:00:00'));
    expect(fakeSerializedDate.toJSON()).toEqual('2019-03-24T00:00:00.000Z');
  });

  it('should transform midday date to fake serialized one', () => {
    const fakeSerializedDate = DateFormatter.toFakeSerializedDate(new Date('2019-03-24T12:00:00'));
    expect(fakeSerializedDate.toJSON()).toEqual('2019-03-24T12:00:00.000Z');
  });

});


projects/angular-stdlib/date/src/lib/date-validators.spec.ts
import { FormControl, ValidatorFn } from '@angular/forms';
import { DateValidators } from './date-validators';

describe('DateValidators', () => {

  const badFormat = new Date('Invalid date');
  const isoString = '2019-01-01';
  const firstDate = new Date('2018-01-01');
  const secondDate = new Date('2020-01-01');

  it('should have valid date', () => {
    const dateControl = new FormControl(new Date(), DateValidators.date);
    expect(dateControl.valid).toBeTruthy();
    expect(dateControl.hasError('date')).toBeFalsy();
  });

  it('should have valid date (null)', () => {
    const dateControl = new FormControl(null, DateValidators.date);
    expect(dateControl.valid).toBeTruthy();
    expect(dateControl.hasError('date')).toBeFalsy();
  });

  it('should have valid date (undefined)', () => {
    const dateControl = new FormControl(undefined, DateValidators.date);
    expect(dateControl.valid).toBeTruthy();
    expect(dateControl.hasError('date')).toBeFalsy();
  });

  it('should have valid date (empty string)', () => {
    const dateControl = new FormControl('', DateValidators.date);
    expect(dateControl.valid).toBeTruthy();
    expect(dateControl.hasError('date')).toBeFalsy();
  });

  it('should have invalid date (bad format)', () => {
    const dateControl = new FormControl(badFormat, DateValidators.date);
    expect(dateControl.valid).toBeFalsy();
    expect(dateControl.hasError('date')).toBeTruthy();
  });

  it('should have invalid date (ISO string)', () => {
    const dateControl = new FormControl(isoString, DateValidators.date);
    expect(dateControl.valid).toBeFalsy();
    expect(dateControl.hasError('date')).toBeTruthy();
  });

  it('should have invalid date (number)', () => {
    const dateControl = new FormControl(0, DateValidators.date);
    expect(dateControl.valid).toBeFalsy();
    expect(dateControl.hasError('date')).toBeTruthy();
  });

  it('should have invalid date (all DateValidators)', () => {
    const dateControl = new FormControl(isoString, [
      DateValidators.date,
      DateValidators.beforeDate(firstDate),
      DateValidators.notBeforeDate(secondDate),
      DateValidators.afterDate(secondDate),
      DateValidators.notAfterDate(firstDate)
    ]);
    expect(dateControl.valid).toBeFalsy();
    expect(dateControl.hasError('date')).toBeTruthy();
    expect(dateControl.hasError('beforeDate')).toBeFalsy();
    expect(dateControl.hasError('notBeforeDate')).toBeFalsy();
    expect(dateControl.hasError('afterDate')).toBeFalsy();
    expect(dateControl.hasError('notAfterDate')).toBeFalsy();
  });

  it('should be before date valid', () => {
    expectValidDate(firstDate, DateValidators.beforeDate(secondDate), 'beforeDate');
  });

  it('should be before date valid (bad format)', () => {
    expectValidDate(badFormat, DateValidators.beforeDate(secondDate), 'beforeDate');
  });

  it('should be before date valid (ISO string)', () => {
    expectValidDate(isoString, DateValidators.beforeDate(secondDate), 'beforeDate');
  });

  it('should be before date valid (null param)', () => {
    expectValidDate(new Date(), DateValidators.beforeDate(null), 'beforeDate');
  });

  it('should be before date invalid (same dates)', () => {
    const dateControl = new FormControl(firstDate, DateValidators.beforeDate(firstDate));
    expect(dateControl.valid).toBeFalsy();
    expect(dateControl.hasError('beforeDate')).toBeTruthy();
  });

  it('should be before date invalid (date after)', () => {
    const dateControl = new FormControl(secondDate, DateValidators.beforeDate(firstDate));
    expect(dateControl.valid).toBeFalsy();
    expect(dateControl.getError('beforeDate').dateName).toBeUndefined();
  });

  it('should be before date invalid with dateName', () => {
    const dateControl = new FormControl(secondDate, DateValidators.beforeDate(firstDate, 'birthdate'));
    expect(dateControl.valid).toBeFalsy();
    expect(dateControl.getError('beforeDate').dateName).toBe('birthdate');
  });

  it('should not be before date valid', () => {
    expectValidDate(secondDate, DateValidators.notBeforeDate(firstDate), 'notBeforeDate');
  });

  it('should not be before date valid (bad format)', () => {
    expectValidDate(badFormat, DateValidators.notBeforeDate(secondDate), 'notBeforeDate');
  });

  it('should not be before date valid (ISO string)', () => {
    expectValidDate(isoString, DateValidators.notBeforeDate(secondDate), 'notBeforeDate');
  });

  it('should not be before date valid (null param)', () => {
    expectValidDate(new Date(), DateValidators.notBeforeDate(null), 'notBeforeDate');
  });

  it('should not be before date valid (same dates)', () => {
    expectValidDate(firstDate, DateValidators.notBeforeDate(firstDate), 'notBeforeDate');
  });

  it('should not be before date invalid (date before)', () => {
    const dateControl = new FormControl(firstDate, DateValidators.notBeforeDate(secondDate));
    expect(dateControl.valid).toBeFalsy();
    expect(dateControl.getError('notBeforeDate').dateName).toBeUndefined();
  });

  it('should not be before date invalid with dateName', () => {
    const dateControl = new FormControl(firstDate, DateValidators.notBeforeDate(secondDate, 'birthdate'));
    expect(dateControl.valid).toBeFalsy();
    expect(dateControl.getError('notBeforeDate').dateName).toBe('birthdate');
  });

  it('should be after date valid', () => {
    expectValidDate(secondDate, DateValidators.afterDate(firstDate), 'afterDate');
  });

  it('should be after date valid (bad format)', () => {
    expectValidDate(badFormat, DateValidators.afterDate(secondDate), 'afterDate');
  });

  it('should be after date valid (ISO string)', () => {
    expectValidDate(isoString, DateValidators.afterDate(secondDate), 'afterDate');
  });

  it('should be after date valid (null param)', () => {
    expectValidDate(new Date(), DateValidators.afterDate(null), 'afterDate');
  });

  it('should be after date invalid (same dates)', () => {
    const dateControl = new FormControl(firstDate, DateValidators.afterDate(firstDate));
    expect(dateControl.valid).toBeFalsy();
    expect(dateControl.hasError('afterDate')).toBeTruthy();
  });

  it('should be after date invalid (date before)', () => {
    const dateControl = new FormControl(firstDate, DateValidators.afterDate(secondDate));
    expect(dateControl.valid).toBeFalsy();
    expect(dateControl.getError('afterDate').dateName).toBeUndefined();
  });

  it('should be after date invalid with dateName', () => {
    const dateControl = new FormControl(firstDate, DateValidators.afterDate(secondDate, 'birthdate'));
    expect(dateControl.valid).toBeFalsy();
    expect(dateControl.getError('afterDate').dateName).toBe('birthdate');
  });

  it('should not be after date valid', () => {
    expectValidDate(firstDate, DateValidators.notAfterDate(secondDate), 'notAfterDate');
  });

  it('should not be after date valid (bad format)', () => {
    expectValidDate(badFormat, DateValidators.notAfterDate(secondDate), 'notAfterDate');
  });

  it('should not be after date valid (ISO string)', () => {
    expectValidDate(isoString, DateValidators.notAfterDate(secondDate), 'notAfterDate');
  });

  it('should not be after date valid (null param)', () => {
    expectValidDate(new Date(), DateValidators.notAfterDate(null), 'notAfterDate');
  });

  it('should not be after date valid (same dates)', () => {
    expectValidDate(firstDate, DateValidators.notAfterDate(firstDate), 'notAfterDate');
  });

  it('should not be after date invalid (date after)', () => {
    const dateControl = new FormControl(secondDate, DateValidators.notAfterDate(firstDate));
    expect(dateControl.valid).toBeFalsy();
    expect(dateControl.getError('notAfterDate').dateName).toBeUndefined();
  });

  it('should not be after date invalid with dateName', () => {
    const dateControl = new FormControl(secondDate, DateValidators.notAfterDate(firstDate, 'birthdate'));
    expect(dateControl.valid).toBeFalsy();
    expect(dateControl.getError('notAfterDate').dateName).toBe('birthdate');
  });

});

function expectValidDate(value: any, validator: ValidatorFn, errorCode: string) {
  const dateControl = new FormControl(value, validator);
  expect(dateControl.valid).toBeTruthy();
  expect(dateControl.hasError(errorCode)).toBeFalsy();
}


projects/angular-stdlib/date/src/lib/date-http-interceptor.spec.ts
import { HTTP_INTERCEPTORS, HttpClient } from '@angular/common/http';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { TestBed } from '@angular/core/testing';
import { DateHttpInterceptor } from './date-http-interceptor';

describe('DateHttpInterceptor', () => {

  let http: HttpClient;
  let mock: HttpTestingController;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [ HttpClientTestingModule ],
      providers: [{
        provide: HTTP_INTERCEPTORS,
        useClass: DateHttpInterceptor,
        multi: true
      }]
    });
    http = TestBed.get(HttpClient);
    mock = TestBed.get(HttpTestingController);
  });

  it('should create an instance', () => {
    expect(new DateHttpInterceptor()).toBeTruthy();
  });

  it('should intercept http request with a null body', () => {
    http.get<any>('/api').subscribe(
      response => expect(response).toBeNull(),
      fail
    );
    verify(mock, null);
  });

  it('should intercept http request without an object body', () => {
    http.get<any>('/api').subscribe(
      response => expect(response).toBe(123),
      fail
    );
    verify(mock, 123);
  });

  it('should intercept http request with a null date', () => {
    http.get<any>('/api').subscribe(
      response => expect(response.date).toBeNull(),
      fail
    );
    verify(mock, { date: null });
  });

  it('should intercept http request with a complex object containing dates', () => {
    http.get<any>('/api').subscribe(
      response => {
        expect(response.firstname).toEqual('Julien');
        expect(response.firstDate).toEqual(new Date('2018-05-15'));
        expect(response.multiDates[0]).toEqual(new Date('2019-09-24'));
        expect(response.multiDates[1]).toEqual(new Date('2017-10-05'));
        expect(response.child.firstname).toEqual('Guillaume');
        expect(response.child.firstDate).toEqual(new Date('2019-02-15'));
        expect(response.child.grandchildren[0].firstname).toEqual('Olivier');
        expect(response.child.grandchildren[0].birthDate).toEqual(new Date('2019-02-18'));
        expect(response.child.grandchildren[0].anotherField).toEqual('BigBoss');
        expect(response.child.grandchildren[1].firstname).toEqual('Mehdi');
        expect(response.child.grandchildren[1].birthDate).toEqual(new Date('1915-01-23'));
        expect(response.child.grandchildren[1].anotherField).toEqual('WhatThe?');
      },
      fail
    );
    verify(mock, {
      firstname: 'Julien',
      firstDate: '2018-05-15',
      multiDates: ['2019-09-24', '2017-10-05'],
      child: {
        firstname: 'Guillaume',
        firstDate: '2019-02-15',
        grandchildren: [{
          firstname: 'Olivier',
          birthDate: '2019-02-18',
          anotherField: 'BigBoss'
        }, {
          firstname: 'Mehdi',
          birthDate: '1915-01-23',
          anotherField: 'WhatThe?'
        }]
      }
    });
  });

  it('should intercept http request with a body containing a date "yyyy-MM-dd"', () => {
    verifyDate(http, mock, '2018-05-15');
  });

  it('should intercept http request with a body containing a datetime "yyyy-MM-ddThh:mm:ss"', () => {
    verifyDate(http, mock, '2018-05-15T15:12:10');
  });

  it('should intercept http request with a body containing a datetime "yyyy-MM-ddThh:mm:ss+00:00"', () => {
    verifyDate(http, mock, '2018-05-15T15:12:10+05:00');
  });

  it('should intercept http request with a body containing a datetime "yyyy-MM-ddThh:mm:ss-01:00"', () => {
    verifyDate(http, mock, '2018-05-15T15:12:10-05:00');
  });

  it('should intercept http request with a body containing a datetime "yyyy-MM-ddThh:mm:ssZ"', () => {
    verifyDate(http, mock, '2018-05-15T15:12:10Z');
  });

  it('should intercept http request with a body containing a datetime "yyyy-MM-ddThh:mm:ss.sss"', () => {
    verifyDate(http, mock, '2018-05-15T15:12:10.524');
  });

  it('should intercept http request with a body containing a datetime "yyyy-MM-ddThh:mm:ss.sss+00:00"', () => {
    verifyDate(http, mock, '2018-05-15T15:12:10.524+05:00');
  });

  it('should intercept http request with a body containing a datetime "yyyy-MM-ddThh:mm:ss.sss-01:00"', () => {
    verifyDate(http, mock, '2018-05-15T15:12:10.524-05:00');
  });

  it('should intercept http request with a body containing a datetime "yyyy-MM-ddThh:mm:ss.sssZ"', () => {
    verifyDate(http, mock, '2018-05-15T15:12:10.524Z');
  });

});

function verify(mock: HttpTestingController, body: any) {
  const request = mock.expectOne(req => req.method === 'GET' && req.url === '/api');
  request.flush(body);
  mock.verify();
}

function verifyDate(http: HttpClient, mock: HttpTestingController, dateAsString: string) {
  http.get<any>('/api').subscribe(
    response => expect(response.date).toEqual(new Date(dateAsString)),
    fail
  );
  verify(mock, { date: dateAsString });
}


projects/angular-stdlib/date/karma.conf.js
// Karma configuration file, see link for more information
// https://karma-runner.github.io/1.0/config/configuration-file.html

module.exports = function (config) {
  config.set({
    basePath: '',
    frameworks: ['jasmine', '@angular-devkit/build-angular'],
    plugins: [
      require('karma-jasmine'),
      require('karma-chrome-launcher'),
      require('karma-jasmine-html-reporter'),
      require('karma-coverage-istanbul-reporter'),
      require('@angular-devkit/build-angular/plugins/karma')
    ],
    client: {
      clearContext: false // leave Jasmine Spec Runner output visible in browser
    },
    coverageIstanbulReporter: {
      dir: require('path').join(__dirname, '../../../coverage/angular-stdlib/date'),
      reports: ['html', 'lcovonly'],
      fixWebpackSourcePaths: true
    },
    reporters: ['progress', 'kjhtml'],
    port: 9876,
    colors: true,
    logLevel: config.LOG_INFO,
    autoWatch: true,
    browsers: ['Chrome'],
    singleRun: false,
    restartOnFileChange: true
  });
};


projects/angular-stdlib/date/src/lib/date-formatter.ts
export class DateFormatter {

  static readonly ISO_DATE = 'yyyy-MM-dd';

  /**
   * Transform the date in the same way as it is serialized :
   * Serialization use JSON.stringify method which use the Date toJSON() method that transform date to UTC Z.
   * Only use when you have to serialize dates and you don't want than serialization apply zero UTC offset transformation.
   *
   * https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/JSON/stringify
   * https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/Date/toJSON
   * https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/Date/toISOString
   *
   * @param date Date to transform
   */
  public static toFakeSerializedDate(date: Date): Date {
    if (date === null || date === undefined) {
      return date;
    }
    const hourAsMinutes = 60;
    const timezoneOffset = date.getTimezoneOffset();

    const convertedDate = new Date(date);
    convertedDate.setHours(date.getHours() - (timezoneOffset / hourAsMinutes));
    convertedDate.setMinutes(date.getMinutes() - (timezoneOffset % hourAsMinutes));
    return convertedDate;
  }

}


projects/angular-stdlib/date/src/lib/date-validators.ts
import { AbstractControl, ValidatorFn, ValidationErrors } from '@angular/forms';

function isEmptyValue(value: any): boolean {
  return value == null || value.length === 0;
}

function isValid(value: any, dateParam: Date, operator): boolean {
  if (isEmptyValue(value) || !(value instanceof Date) || isNaN(value.getTime()) || dateParam == null || operator(value, dateParam)) {
    return true;
  }
  return false;
}

function buildErrorValue(dateName?: string): true | { dateName: string } {
  return dateName ? { dateName } : true;
}

function date(control: AbstractControl): ValidationErrors | null {
  const value = control.value;
  if (isEmptyValue(value) || (value instanceof Date && !isNaN(value.getTime()))) {
    return null;
  }
  return { date: true };
}

function beforeDate(dateParam: Date, dateName?: string): ValidatorFn {
  return (control: AbstractControl) => {
    return isValid(control.value, dateParam, (a, b) => a < b) ? null : { beforeDate: buildErrorValue(dateName) };
  };
}

function notBeforeDate(dateParam: Date, dateName?: string): ValidatorFn {
  return (control: AbstractControl) => {
    return isValid(control.value, dateParam, (a, b) => a >= b) ? null : { notBeforeDate: buildErrorValue(dateName) };
  };
}

function afterDate(dateParam: Date, dateName?: string): ValidatorFn {
  return (control: AbstractControl) => {
    return isValid(control.value, dateParam, (a, b) => a > b) ? null : { afterDate: buildErrorValue(dateName) };
  };
}

function notAfterDate(dateParam: Date, dateName?: string): ValidatorFn {
  return (control: AbstractControl) => {
    return isValid(control.value, dateParam, (a, b) => a <= b) ? null : { notAfterDate: buildErrorValue(dateName) };
  };
}

/**
 * @description
 * Provides a set of built-in validators for Dates that can be used by form controls.
 *
 * A validator is a function that processes a `FormControl` or collection of
 * controls and returns an error map or null. A null map means that validation has passed.
 *
 * @publicApi
 */
export const DateValidators = {
  /**
   * @description
   * Validator that requires the control's value to be a `Date`
   *
   * @usageNotes Usage :
   *
   * ```typescript
   * const control = new FormControl(123, DateValidators.date);
   * console.log(control.errors); // {date: true}
   * ```
   *
   * @returns An error map with the `date` property if the validation check fails, otherwise `null`.
   */
  date,

  /**
   * @description
   * Validator that requires the control's value to be prior than the date in parameter.
   *
   * @usageNotes Usage :
   * ```typescript
   * const control = new FormControl(new Date('2020-01-01'), DateValidators.beforeDate(new Date('2019-01-01')));
   * console.log(control.errors); // {beforeDate: true}
   * ```
   * With optional `dateName` parameter :
   * ```typescript
   * const control = new FormControl(new Date('2020-01-01'), DateValidators.beforeDate(new Date('2019-01-01'), 'birthdate'));
   * console.log(control.errors); // {beforeDate: {dateName: 'birthdate'}}
   * ```
   *
   * @returns A validator function that returns an error map with the
   * `beforeDate` property if the validation check fails, otherwise `null`.
   */
  beforeDate,

  /**
   * @description
   * Validator that requires the control's value not to be prior than the date in parameter.
   *
   * @usageNotes Usage :
   * ```typescript
   * const control = new FormControl(new Date('2019-01-01'), DateValidators.notBeforeDate(new Date('2020-01-01')));
   * console.log(control.errors); // {notBeforeDate: true}
   * ```
   * With optional `dateName` parameter :
   * ```typescript
   * const control = new FormControl(new Date('2019-01-01'), DateValidators.notBeforeDate(new Date('2020-01-01'), 'birthdate'));
   * console.log(control.errors); // {notBeforeDate: {dateName: 'birthdate'}}
   * ```
   *
   * @returns A validator function that returns an error map with the
   * `notBeforeDate` property if the validation check fails, otherwise `null`.
   */
  notBeforeDate,

  /**
   * @description
   * Validator that requires the control's value to be later than the date in parameter.
   *
   * @usageNotes Usage :
   * ```typescript
   * const control = new FormControl(new Date('2019-01-01'), DateValidators.afterDate(new Date('2020-01-01')));
   * console.log(control.errors); // {afterDate: true}
   * ```
   * With optional `dateName` parameter :
   * ```typescript
   * const control = new FormControl(new Date('2019-01-01'), DateValidators.afterDate(new Date('2020-01-01'), 'birthdate'));
   * console.log(control.errors); // {afterDate: {dateName: 'birthdate'}}
   * ```
   *
   * @returns A validator function that returns an error map with the
   * `afterDate` property if the validation check fails, otherwise `null`.
   */
  afterDate,

  /**
   * @description
   * Validator that requires the control's value not to be later than the date in parameter.
   *
   * @usageNotes Usage :
   * ```typescript
   * const control = new FormControl(new Date('2020-01-01'), DateValidators.notAfterDate(new Date('2019-01-01')));
   * console.log(control.errors); // {notAfterDate: true}
   * ```
   * With optional `dateName` parameter :
   * ```typescript
   * const control = new FormControl(new Date('2020-01-01'), DateValidators.notAfterDate(new Date('2019-01-01'), 'birthdate'));
   * console.log(control.errors); // {notAfterDate: {dateName: 'birthdate'}}
   * ```
   *
   * @returns A validator function that returns an error map with the
   * `notAfterDate` property if the validation check fails, otherwise `null`.
   */
  notAfterDate
};


projects/angular-stdlib/date/src/lib/date-http-interceptor.ts
import { Injectable } from '@angular/core';
import { HttpRequest, HttpHandler, HttpEvent, HttpInterceptor, HttpResponse } from '@angular/common/http';
import { Observable } from 'rxjs';
import { tap } from 'rxjs/operators';

/**
 * Interceptor that handle Http Request and convert from JSON response body : ISO 8601 date and datetime to Date object.
 * Only necessary due to ngx-bootstrap/datepicker dependency use, this one does not handle "ISO string dates" unlike Angular.
 * Linked to this unresolved issue : https://github.com/valor-software/ngx-bootstrap/issues/4487
 */
@Injectable()
export class DateHttpInterceptor implements HttpInterceptor {

  /**
   * Regular expression for ISO 8601 date and date time
   * Accepted formats :
   * - 'yyyy-MM-dd'
   * - 'yyyy-MM-ddThh:mm:ss'
   * - 'yyyy-MM-ddThh:mm:ss+00:00'
   * - 'yyyy-MM-ddThh:mm:ss-01:00'
   * - 'yyyy-MM-ddThh:mm:ssZ'
   * - 'yyyy-MM-ddThh:mm:ss.sss'
   * - 'yyyy-MM-ddThh:mm:ss.sss+00:00'
   * - 'yyyy-MM-ddThh:mm:ss.sss-01:00'
   * - 'yyyy-MM-ddThh:mm:ss.sssZ'
   */
  private readonly iso8601RegExp = /^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2}(\.\d+)?(([+-]\d{2}:\d{2})|Z)?)?$/;

  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    return next.handle(req).pipe(
      tap((event: HttpEvent<any>) => {
        if (event instanceof HttpResponse) {
          this.convertToDate(event.body);
        }
      })
    );
  }

  private convertToDate(body) {
    if (body === null || body === undefined || typeof body !== 'object') {
      return;
    }
    for (const key of Object.keys(body)) {
      const value = body[key];
      if (this.isIso8601(value)) {
        body[key] = new Date(value);
      } else if (typeof value === 'object') {
        this.convertToDate(value);
      }
    }
  }

  private isIso8601(value) {
    if (value === null || value === undefined) {
      return false;
    }
    return this.iso8601RegExp.test(value);
  }

}


projects/angular-stdlib/date/src/lib/angular-date-http-interceptor.ts
import { Injectable } from '@angular/core';
import { HttpRequest, HttpHandler, HttpEvent, HttpInterceptor, HttpResponse } from '@angular/common/http';
import { Observable } from 'rxjs';
import { tap } from 'rxjs/operators';

/**
 * Interceptor that handle Http Request and convert from JSON response body : ISO 8601 date and datetime to Date object.
 * Only necessary due to ngx-bootstrap/datepicker dependency use, this one does not handle "ISO string dates" unlike Angular.
 * Linked to this unresolved issue : https://github.com/valor-software/ngx-bootstrap/issues/4487
 */
@Injectable()
export class AngularDateHttpInterceptor implements HttpInterceptor {

  /**
   * Regular expression for ISO 8601 date and date time
   * Accepted formats :
   * - 'yyyy-MM-dd'
   * - 'yyyy-MM-ddThh:mm:ss'
   * - 'yyyy-MM-ddThh:mm:ss+00:00'
   * - 'yyyy-MM-ddThh:mm:ss-01:00'
   * - 'yyyy-MM-ddThh:mm:ssZ'
   * - 'yyyy-MM-ddThh:mm:ss.sss'
   * - 'yyyy-MM-ddThh:mm:ss.sss+00:00'
   * - 'yyyy-MM-ddThh:mm:ss.sss-01:00'
   * - 'yyyy-MM-ddThh:mm:ss.sssZ'
   */
  private readonly iso8601RegExp = /^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2}(\.\d+)?(([+-]\d{2}:\d{2})|Z)?)?$/;

  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    return next.handle(req).pipe(
      tap((event: HttpEvent<any>) => {
        if (event instanceof HttpResponse) {
          this.convertToDate(event.body);
        }
      })
    );
  }

  private convertToDate(body) {
    if (body === null || body === undefined || typeof body !== 'object') {
      return;
    }
    for (const key of Object.keys(body)) {
      const value = body[key];
      if (this.isIso8601(value)) {
        body[key] = new Date(value);
      } else if (typeof value === 'object') {
        this.convertToDate(value);
      }
    }
  }

  private isIso8601(value) {
    if (value === null || value === undefined) {
      return false;
    }
    return this.iso8601RegExp.test(value);
  }

}
